Switch our main macOS actions to Arm.

chandlerc opened the pull request:
GitHub has publicly available Arm macOS runners now: https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

These runners are used when the OS label is `macos-14` instead of `macos-12`.

This PR switches the main PR and merge queue macOS runs to that OS. It adds a push-only `macos-12` run to check post-merge whether Intel macOS keeps working as we don't want that to break too badly. Given how much faster the M1 is, it didn't seem worth it to keep this in the critical path of development. And the Intel path here is more likely than others to diminish in importance over time.

All our OS predicates in the action encoded a specific OS version which no longer works given two macOS versions, so I've switch all the OS predicates to a generic pattern that doesn't encode a version.

Example run on a PR: https://github.com/carbon-language/carbon-lang/actions/runs/7766629200?pr=3686
Example run on push: https://github.com/carbon-language/carbon-lang/actions/runs/7766480822

chandlerc committed:
Switch our main macOS actions to Arm.

GitHub has publicly available Arm macOS runners now:
https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

These runners are used when the OS label is `macos-14` instead of
`macos-12`.

This PR switches the main PR and merge queue macOS runs to that OS. It
adds a push-only `macos-12` run to check post-merge whether Intel macOS
keeps working as we don't want that to break too badly. Given how much
faster the M1 is, it didn't seem worth it to keep this in the critical
path of development. And the Intel path here is more likely than others
to diminish in importance over time.

All our OS predicates in the action encoded a specific OS version which
no longer works given two macOS versions, so I've switch all the OS
predicates to a generic pattern that doesn't encode a version.

Example run on our `actions-test` branch: https://github.com/carbon-language/carbon-lang/actions/runs/7740538377

Note that before landing this, we'll need to turn off the requirement
for macos-12 to test actions on our merge queue, and then after this
runs we'll need to turn back on macos-14.

We can try to change how we name things to avoid this kind of churn if
folks prefer.

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -23,19 +23,26 @@ jobs:
             build_mode: 'clang-tidy'
+          # On push (post-submit), also run on `macos-12` to get Intel macOS
+          # coverage. This is less and less critical so we skip it on the PR and
+          # merge queue path by switching to a redundant OS.
```
Just to be sure, is this still important enough that it's worth the additional costs involved?

chandlerc replied:
I think it's nice to keep for now. If nothing else, the fact that so much GitHub action infra is still using Intel macOS seems like a good reason to avoid letting it break just yet.

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -23,19 +23,26 @@ jobs:
+          # On push (post-submit), also run on `macos-12` to get Intel macOS
+          # coverage. This is less and less critical so we skip it on the PR and
+          # merge queue path by switching to a redundant OS.
+          - os: ${{ github.event_name == 'push' && 'macos-12' || 'macos-14' }}
```
I'm trusting this works, but had you considered on line 26 instead writing:

```
- os: ${{ github.event_name == 'push' && [ubuntu-22.04, macos-14, macos-12] || [ubuntu-22.04, macos-14] }}
```

(if this works, I think it'd reduce repetition)

chandlerc replied:
It's a bit harder to do than that because you can't put JSON directly into an expression, but you can parse a string into JSON, so I've done that. It does seem simpler, even if it breaks VSCode's YAML highlighting...

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -74,33 +81,43 @@ jobs:
           # Match the min version listed in docs/project/contribution_tools.md
-          python-version: '3.9'
+          # or the oldest version available on the OS.
+          python-version: ${{ matrix.os == 'macos-14' && '3.11' || '3.9' }}
```
Had you considered updating the min version? I don't think we've changed for a couple years now, maybe updating it to 3.11 everywhere would be fine?

chandlerc replied:
Happy to try that, but seemed like maybe better to do separately to minimize how much changes with this PR?

jonmeow replied:
SGTM

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -23,19 +23,26 @@ jobs:
       # comparison, MacOS images have >100GB free.
       - name: Free up disk space (Ubuntu)
-        if: matrix.os == 'ubuntu-22.04'
+        if: contains(matrix.os, 'ubuntu')
```
Why contains instead of startsWith?

chandlerc replied:
It was shorter, but maybe that's not a great reason.

That said, this still was really repetitive. I've extracted a job-level environment context variable `os` to just contain `macos` or `ubuntu`, and renamed the matrix variable to `runner`. This lets us simply use `runner` when we want a *specific* runner image, and `os` when we just need to know whether its macOS or Ubuntu.

PTAL and let me know what you think. Happy to go back or try different approaches here.

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -74,33 +81,43 @@ jobs:
           cache-name: cache-homebrew
         with:
-          # Cover all the critical parts of Homebrew here.
+          # Cover all the critical parts of Homebrew here. Homebrew on Arm macOS
+          # uses its own prefix making this easy to cover, but we need a few
+          # different paths for Intel.
```
Just to be sure, this is arm rather than the macos-14 setup?

chandlerc replied:
Yep, all of Arm Homebrew uses this layout.

My guess is they made the change to a parallel root because there was a pretty long period of time where the *x86* Homebrew was in use on Arm machines via Rosetta. This "just worked" for a bunch of use cases, and there were a lot of installs like this. When they got ready to have a native Arm Homebrew, they wanted it to be able to be installed in parallel with the x86 one to avoid breakage, so they moved to their own prefix.

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -227,12 +248,15 @@ jobs:
       # We need to replace the `.` with a `_` for the build cache.
       - name: Setup LLVM and Clang (macOS)
-        if: steps.filter.outputs.has_code == 'true' && matrix.os == 'macos-12'
+        if:
+          steps.filter.outputs.has_code == 'true' && contains(matrix.os,
+          'macos')
```
Why are you using macos-12 for the macos-14 cache? Will the build outputs all be the same, even with the environment and CPU change?

chandlerc replied:
No, good catch. I had missed that the version was included here. See below for a full answer.

jonmeow reviewed (commented):
Generally looks good, just a few questions about structure.

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -227,12 +248,15 @@ jobs:
       - name: Setup LLVM and Clang (Ubuntu)
         if:
-          steps.filter.outputs.has_code == 'true' && matrix.os == 'ubuntu-22.04'
+          steps.filter.outputs.has_code == 'true' && contains(matrix.os,
+          'ubuntu')
```
Similar to the macos-12 comment, are you sure all ubuntu builds are sufficiently hermetic that they will yield the same outputs regardless of version? Note, I think we've seen some corrupt builds from caching of non-hermetic build outputs before -- IIRC that's https://github.com/carbon-language/carbon-lang/pull/2198 and its note about a slight difference in include paths.

jonmeow replied:
Note, maybe if the build hermeticism issue is solved, we could just remove the os suffix now?

chandlerc replied:
No, they're not. Adding ARM works "fine", but by accident -- the *CPU* part of this is part of the action hash used to query the cache. But the OS versions are not. Nor are the distinctions between Ubuntu and some other Linux distro if we ever care.

But there *is* a great way to solve the hermeticity issues here more simply than partitioning our buckets! We can use the `--action_env` flag to thread a key for the remote cache in from the workflow file, and set that in a way that allows us to use a single bucket.

I've done this to make it more obvious that the two macOS runs are correct, and left a TODO to completely remove this -- that seems like it's worth keeping as a separate change as we'll need to fix up the name of the buckets, the keys, and then do a round of fresh builds to re-populate them.

jonmeow replied:
Can separate out if you want, but I've created a carbon-builds-github-v1. (just dropping the os suffix)

> that seems like it's worth keeping as a separate change as we'll need to fix up the name of the buckets, the keys, and then do a round of fresh builds to re-populate them.

I think all you'd need to do is change the names of the bucket in this PR. Fresh builds are also necessary with either the macos-14 or the --action_env addition (both require cache population), you'd just be putting builds into a different bucket.

Keys are unaffected since it's still the same account.

chandlerc replied:
Ah cool, nice that it's that easy. I've done this and tested that it all works, and we get caching afterward.

chandlerc committed:
Restructure based on code review.

This redoes the matrix setup based on a review suggestion, encapsulates
the OS testing in the environment context, and adds a Bazel-level
caching key to avoid colliding the cache between different OS versions.

chandlerc reviewed (commented):
Thanks for the great feedback! Addressing it resulted in quite some changes, so PTAL.

One run using the new techniques is here: https://github.com/carbon-language/carbon-lang/actions/runs/7753457932

This had to populate the now-keyed remote caches. I can push some fresh versions to double check that the cache hits work.

jonmeow commented on .github/workflows/tests.yaml:
```diff
@@ -358,6 +381,7 @@ jobs:
 
             bazel_exit=0
             bazelisk test -c ${{ matrix.build_mode }} \
+              --action_env=BAZEL_REMOTE_CACHE_KEY \
```
Can this be added to the bazelrc instead of being passed to individual executions, to ensure it also applies even if `bazel mod deps` or the commands executed as part of diff do cache reads/writes?

chandlerc replied:
Doh, yes. Not sure why I didn't think of that, thanks.

jonmeow commented:
Feel free to merge when you've had a chance to respond on comments, but I do wonder if action_env should be in the bazelrc. I'm glad you found that though.

chandlerc committed:
simplify based on review

chandlerc committed:
test change

chandlerc committed:
undo

chandlerc reviewed (commented):
I think this is all ready to go, but I'll merge on Monday in case it creates some problem.

chandlerc committed:
Merge branch 'trunk' into macos-actions
